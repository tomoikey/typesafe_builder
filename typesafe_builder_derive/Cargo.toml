[package]
name = "typesafe_builder_derive"
version = "1.3.0"
edition = "2024"
authors = ["tomoikey"]
readme = "README.md"
repository = "https://github.com/tomoikey/typesafe_builder"
categories = ["accessibility", "development-tools", "rust-patterns"]
license = "MIT"
description = "A procedural macro to generate type-safe builder patterns for Rust structs"

[lib]
proc-macro = true

[dependencies]
darling = "0.20.11"
proc-macro2 = "1.0.95"
syn = { version = "2.0.101", features = ["derive", "parsing", "proc-macro"] }
quote = "1.0.40"

[dev-dependencies]
trybuild = "1.0.105"
