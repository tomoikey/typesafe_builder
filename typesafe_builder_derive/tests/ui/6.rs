use typesafe_builder_derive::Builder;

struct _TypesafeBuilderEmpty;
struct _TypesafeBuilderFilled;

fn main() {
    #[derive(Builder)]
    struct User {
        #[builder(optional)]
        name: Option<String>,
        #[builder(optional)]
        age: Option<u8>,
        #[builder(optional)]
        address: Option<String>,
        #[builder(required_if = "name && (age || address)")]
        email: Option<String>,
    }

    // compile error because email is required if name and age are present
    let user = UserBuilder::new()
        .with_name("Alice".to_string())
        .with_age(20)
        .build();

    // compile error because email is required if name and address are present
    let user = UserBuilder::new()
        .with_name("Alice".to_string())
        .with_address("123 Main St".to_string())
        .build();
} 