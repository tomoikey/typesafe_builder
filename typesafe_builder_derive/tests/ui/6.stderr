error[E0599]: no method named `build` found for struct `UserBuilder<_TypesafeBuilderFilled, _TypesafeBuilderFilled, _TypesafeBuilderEmpty, _TypesafeBuilderEmpty>` in the current scope
  --> tests/ui/6.rs:23:10
   |
7  |       #[derive(Builder)]
   |                ------- method `build` not found for this struct
...
20 |       let user = UserBuilder::new()
   |  ________________-
21 | |         .with_name("Alice".to_string())
22 | |         .with_age(20)
23 | |         .build();
   | |         -^^^^^ method not found in `UserBuilder<_TypesafeBuilderFilled, _TypesafeBuilderFilled, _TypesafeBuilderEmpty, _TypesafeBuilderEmpty>`
   | |_________|
   |
   |
   = note: the method was found for
           - `UserBuilder<_TypesafeBuilderEmpty, _TypesafeBuilderEmpty, _TypesafeBuilderEmpty, _TypesafeBuilderEmpty>`
           - `UserBuilder<_TypesafeBuilderEmpty, _TypesafeBuilderEmpty, _TypesafeBuilderEmpty, _TypesafeBuilderFilled>`
           - `UserBuilder<_TypesafeBuilderEmpty, _TypesafeBuilderEmpty, _TypesafeBuilderFilled, _TypesafeBuilderEmpty>`
           - `UserBuilder<_TypesafeBuilderEmpty, _TypesafeBuilderEmpty, _TypesafeBuilderFilled, _TypesafeBuilderFilled>`
           and 9 more types

error[E0599]: no method named `build` found for struct `UserBuilder<_TypesafeBuilderFilled, _TypesafeBuilderEmpty, _TypesafeBuilderFilled, _TypesafeBuilderEmpty>` in the current scope
  --> tests/ui/6.rs:29:10
   |
7  |       #[derive(Builder)]
   |                ------- method `build` not found for this struct
...
26 |       let user = UserBuilder::new()
   |  ________________-
27 | |         .with_name("Alice".to_string())
28 | |         .with_address("123 Main St".to_string())
29 | |         .build();
   | |         -^^^^^ method not found in `UserBuilder<_TypesafeBuilderFilled, _TypesafeBuilderEmpty, _TypesafeBuilderFilled, _TypesafeBuilderEmpty>`
   | |_________|
   |
   |
   = note: the method was found for
           - `UserBuilder<_TypesafeBuilderEmpty, _TypesafeBuilderEmpty, _TypesafeBuilderEmpty, _TypesafeBuilderEmpty>`
           - `UserBuilder<_TypesafeBuilderEmpty, _TypesafeBuilderEmpty, _TypesafeBuilderEmpty, _TypesafeBuilderFilled>`
           - `UserBuilder<_TypesafeBuilderEmpty, _TypesafeBuilderEmpty, _TypesafeBuilderFilled, _TypesafeBuilderEmpty>`
           - `UserBuilder<_TypesafeBuilderEmpty, _TypesafeBuilderEmpty, _TypesafeBuilderFilled, _TypesafeBuilderFilled>`
           and 9 more types
